data/tools/wmllint

```diff
@@ -1073,42 +1073,9 @@ def standard_unit_filter():
 # but any given ability can be associated with only one special note
 # Some notes are handled directly in the global_sanity_check() function
 notepairs = [
-    ("{ABILITY_HEALS}", "{NOTE_HEALS}"),
-    ("{ABILITY_EXTRA_HEAL}", "{NOTE_EXTRA_HEAL}"),
-    ("{ABILITY_UNPOISON}", "{NOTE_UNPOISON}"),
-    ("{ABILITY_CURES}", "{NOTE_CURES}"),
-    ("{ABILITY_REGENERATES}", "{NOTE_REGENERATES}"),
-    ("{ABILITY_SELF_HEAL}", "{NOTE_SELF_HEAL}"),
-    ("{ABILITY_STEADFAST}", "{NOTE_STEADFAST}"),
-    ("{ABILITY_LEADERSHIP}", "{NOTE_LEADERSHIP}"),
-    ("{ABILITY_SKIRMISHER}", "{NOTE_SKIRMISHER}"),
-    ("{ABILITY_ILLUMINATES}", "{NOTE_ILLUMINATES}"),
-    ("{ABILITY_TELEPORT}", "{NOTE_TELEPORT}"),
-    ("{ABILITY_AMBUSH}", "{NOTE_AMBUSH}"),
-    ("{ABILITY_NIGHTSTALK}", "{NOTE_NIGHTSTALK}"),
-    ("{ABILITY_CONCEALMENT}", "{NOTE_CONCEALMENT}"),
-    ("{ABILITY_SUBMERGE}", "{NOTE_SUBMERGE}"),
-    ("{ABILITY_FEEDING}", "{NOTE_FEEDING}"),
-    ("{ABILITY_INSPIRE}", "{NOTE_INSPIRE}"),
-    ("{ABILITY_INITIATIVE}", "{NOTE_INITIATIVE}"),
-    ("{ABILITY_DISTRACT}", "{NOTE_DISTRACT}"),
-    ("{ABILITY_DISENGAGE}", "{NOTE_DISENGAGE}"),
-    ("{ABILITY_FORMATION}", "{NOTE_FORMATION}"),
-    ("{WEAPON_SPECIAL_BERSERK}", "{NOTE_BERSERK}"),
-    ("{WEAPON_SPECIAL_BACKSTAB}", "{NOTE_BACKSTAB}"),
-    ("{WEAPON_SPECIAL_PLAGUE", "{NOTE_PLAGUE}"), # No } deliberately
-    ("{WEAPON_SPECIAL_SLOW}", "{NOTE_SLOW}"),
-    ("{WEAPON_SPECIAL_PETRIFY}", "{NOTE_PETRIFY}"),
-    ("{WEAPON_SPECIAL_MARKSMAN}", "{NOTE_MARKSMAN}"),
-    ("{WEAPON_SPECIAL_MAGICAL}", "{NOTE_MAGICAL}"),
-    ("{WEAPON_SPECIAL_SWARM}", "{NOTE_SWARM}"),
-    ("{WEAPON_SPECIAL_CHARGE}", "{NOTE_CHARGE}"),
-    ("{WEAPON_SPECIAL_DRAIN}", "{NOTE_DRAIN}"),
-    ("{WEAPON_SPECIAL_FIRSTSTRIKE}", "{NOTE_FIRSTSTRIKE}"),
-    ("{WEAPON_SPECIAL_POISON}", "{NOTE_POISON}"),
-    ("{WEAPON_SPECIAL_STUN}", "{NOTE_STUN}"),
-    ("{WEAPON_SPECIAL_SHOCK}", "{NOTE_SHOCK}"),
-    ("{WEAPON_SPECIAL_DAZE}", "{NOTE_DAZE}"),
+    ("movement_type=mounted", "{NOTE_DEFENSE_CAP}"),
+    ("movement_type=undeadspirit", "{NOTE_SPIRIT}"),
+    ("type=arcane", "{NOTE_ARCANE}"),
```

Comment:
Any reason why arcane is not automatically added?

Reply:
Though admittedly that note seems like ancient leftover without any actual purpose.

Reply:
You're asking me why I didn't make these three notes get automatically added?

That's because they're not linked to a weapon special or unit ability. One is linked to a damage type, the other two to movement types. I _could_ make the latter two be added automatically but the former doesn't have any "central" place to put a special note, so I would have to create such a place which is a significantly larger task.

I do plan to do both those things at some point, but not for 1.16.

Reply:
> I do plan to do both those things at some point, but not for 1.16.

Then `{NOTE_ARCANE}` should itself deprecated, and `{SPECIAL_NOTES_ARCANE}`'s deprecation message should be changed because it's currently telling UMC authors to use a deprecated macro.

Reply:
Deprecating `{NOTE_ARCANE}` before its substitute exists doesn't seem like the correct answer…

Reply:
So follow that logic on to whether it's right to deprecate `{SPECIAL_NOTES_ARCANE}` at this point.

Reply:
Well, it was already deprecated from the previous change that _added_ `{NOTE_ARCANE}`, and furthermore, someone using `{SPECIAL_NOTES_ARCANE}` might end up with two "Special Notes" sections, so I don't think leaving it undeprecated is a good idea.

Reply:
So what is a UMC developer meant to do? This conversation keeps reminding me of @shikadiqueen's complaints that UMC devs keep being made to do busy-work because the core devs aren't planning.

Suggestions:
* Make the deprecation note on `{SPECIAL_NOTES_ARCANE}` say "In *version* this special note will be added automatically; there is no reason to change to `{NOTE_ARCANE}` as that will become a no-op in the same version". Change the deprecation level to match that.
* Deprecate neither tag until a long-term solution is available.

Reply:
I don't think either solution is viable. Use of `{NOTE_ARCANE}` is required to fit in with the auto-added notes from weapons and abilities, and the version when it will be replaced is unknown. It might be 1.17, or it might not be. I think the correct action for a UMC developer _is_ to use `{NOTE_ARCANE}`, although using neither is also an option.

Reply:
Adding these three note pairs isn't really that useful, because they are already handled elsewhere in wmllint: https://github.com/wesnoth/wesnoth/blob/f1fdbc81e2a383234f1dd59c85a97383d82a7585/data/tools/wmllint#L1486-L1531.

Reply:
Huh, okay. I added them because it looked like my previous special notes commit had incorrectly removed them, but if they're handled separately, I guess I'll take them out again.

Reply:
> You're asking me why I didn't make these three notes get automatically added?
> 
> That's because they're not linked to a weapon special or unit ability. One is linked to a damage type, the other two to movement types. I _could_ make the latter two be added automatically but the former doesn't have any "central" place to put a special note, so I would have to create such a place which is a significantly larger task.

I'm coming to the conclusion that (doing the full task) would be easier than (handling the deprecation properly with the task half-done). It would also avoid making busy-work for UMC creators.

> version when it will be replaced is unknown. It might be 1.17, or it might not be.

In the same way that opening a PR makes an implied commitment of your own time to progressing it, I feel merging this PR should have an implied commitment to the follow-up task in 1.17. I'm feeling quite invested myself having spent time on the review; if you want me to then I'll take over getting the full task done for 1.15.14.

Reply:
Does that mean you're offering to implement:

* A `special_note` key in `[units][movetype]` (side note: there might be other places it could be supported if you wanted, eg in `[race]`, but that's not required for these specific three notes)
* A `[damage_type]` tag (probably under `[units]`? not sure) that includes `special_note` and also supercedes `[language]type_$TYPEID` as well as specifying things like damage icon and colour

If you want to commit to that, then I'm fine with it.

Reply:
> Huh, okay. I added them because it looked like my previous special notes commit had incorrectly removed them, but if they're handled separately, I guess I'll take them out again.
> 
> Do you think I should remove this whole system as well, or just leave the list empty by default?

For now, just leave the list empty; once this PR is merged, I'll handle its removal, if needed.

Reply:
> Does that mean you're offering to implement:

I think my offer is:
* Add `[units][movetype][special_note]` to replace `NOTE_SPIRIT`.
* Replace `NOTE_DEFENSE_CAP` with something that's auto-added. If we don't agree about where the auto-adding code should go, I'll move the note to `[units][movetype][special_note]` for the first PR and then do a follow-up PR changing it to auto-adding.
* While agreeing that it's a hack, add `[language]special_note_damage_type_arcane` and auto-add `special_note_damage_type_$TYPEID` to any unit with a `TYPEID` attack.

I like your idea of replacing `[language]type=$TYPEID` with a `[damage_type]` tag, but feel it's too big a change to get in to 1.16. While I can see some UMC authors using `[language]special_note_damage_type_electric` etc, it will be far fewer authors and far fewer locations that `NOTE_SPIRIT`; so I'd be happy about adding that for 1.16 and then deprecating it in 1.18.

Reply:
I think that's pretty much an acceptable compromise, but why `[units][movetype][special_note]` instead of [units][movetype]special_note=`?

Reply:
I hadn't thought it over when I posted, but the UTBS units with NOTE_ETHEREAL have two special notes that should probably both move to their movetype.

Reply:
I didn't look into what ethereal really connects to… is it present exactly when (if and only if) a specific movetype is used? If so, then sure.

Reply:
Having looked more at NOTE_ETHEREAL, there isn't a corresponding movetype. It's used in a set of specialised ghosts, which have the base unit's movetype and then some custom movement costs on top.

Reply:
> For now, just leave the list empty; once this PR is merged, I'll handle its removal, if needed.

@Elvish-Hunter: it ended up merging as #5811 instead of #5798, but please would you do the update for wmllint?